### DISCUSSION TITLE
TransformerConv RuntimeError: Not compiled with CUDA support

### DISCUSSION THREAD

**Discussion by roshandhakal** [pyg-team/pytorch_geometric#4245] [Q&A]
I can easily train my model with TransformerConv with my cpu. However, when gpu is enabled, it throws me this error. I don't know if its issue with my machine or TransformerConv itself. Can someone please help? Full traceback of the error: 

`Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.8/dist-packages/envprep/scripts/gcn_train.py", line 164, in <module>
    main(args)
  File "/usr/local/lib/python3.8/dist-packages/envprep/scripts/gcn_train.py", line 63, in main
    out = model.forward(batch, device)
  File "/usr/local/lib/python3.8/dist-packages/envprep/models/prepare_env_gcn.py", line 22, in forward
    x = self.conv1(x, edge_index, edge_attr)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch_geometric/nn/conv/transformer_conv.py", line 176, in forward
    out = self.propagate(edge_index, query=query, key=key, value=value,
  File "/usr/local/lib/python3.8/dist-packages/torch_geometric/nn/conv/message_passing.py", line 317, in propagate
    out = self.message(**msg_kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch_geometric/nn/conv/transformer_conv.py", line 216, in message
    alpha = softmax(alpha, index, ptr, size_i)
RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/torch_scatter/scatter.py", line 160, in softmax
        return scatter_min(src, index, dim, out, dim_size)[0]
    elif reduce == 'max':
        return scatter_max(src, index, dim, out, dim_size)[0]
               ~~~~~~~~~~~ <--- HERE
    else:
        raise ValueError
  File "/usr/local/lib/python3.8/dist-packages/torch_scatter/scatter.py", line 72, in scatter_max
        out: Optional[torch.Tensor] = None,
        dim_size: Optional[int] = None) -> Tuple[torch.Tensor, torch.Tensor]:
    return torch.ops.torch_scatter.scatter_max(src, index, dim, out, dim_size)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: Not compiled with CUDA support

Training Device: cuda`

**Comment by rusty1s** [comment]
It looks like `torch-scatter` is not compiled with CUDA support. Can you share some more information about how you installed `torch-scatter`. Can you try to re-install it via `pip install --verbose` as well?

**Reply by roshandhakal** [reply]
I am installing it inside the docker container. This is how I have installed the modules. RUN `pip3 install torch-scatter torch-sparse torch-cluster torch-spline-conv -f https://pytorch-geometric.com/whl/torch-1.8.0+cu111.html`

Is it because of the cuda version and the way it is installed? I have also an error running GCNConv with GPU. I can run on CPU, however. Looks like its something to do with GPU?

This is the output for nvidia-smi. 
![image](https://user-images.githubusercontent.com/30138013/158063457-b487fa9b-e254-4685-9495-cb21a2727eb7.png)

**Reply by rusty1s** [reply]
Can you try to install via explicit versions, see [here](https://data.pyg.org/whl/torch-1.8.0%2Bcu111.html):
```
pip3 install torch-scatter==2.0.8 torch-sparse==0.6.12 torch-cluster==1.5.9 torch-spline-conv==1.2.1 -f https://pytorch-geometric.com/whl/torch-1.8.0+cu111.html
```

**Reply by roshandhakal** [reply]
Can you tell me why this specific version?

**Reply by roshandhakal** [reply]
I thing I noticed strange was; I trained the model on gpu and generated a network file after training. Now, I want to use the file, but I get the error on same thing. Now, the training has the error too. I am wondering if its a error caused by gpu?

**Reply by roshandhakal** [reply]
I am getting this error with explicit versions: 
![image](https://user-images.githubusercontent.com/30138013/158389574-1316ffac-e060-4698-942e-184f96930649.png)

**Reply by rusty1s** [reply]
Is the previous issue resolved? For the second one, it would be good to have some minimal reproducible script.

If you are using `GCNConv`, ensure that any edge weights are positive, as otherwise the output of it may include NaNs.

**Reply by roshandhakal** [reply]
I am sorry for my delayed response. I was away due to spring break. The thing is everything worked fine earlier, both TransformerConv and GCNConv with gpu, no issues, not even error I showed in the discussion. Suddenly, it started showing all these errors with gpu. However, it works with cpu. Both TransformerConv and GCNConv. I honestly don't know how to go about this. 

For GCNConv, there are no edge weights anyway. And, no previous issue with TransformerConv is not resolved.

**Reply by rusty1s** [reply]
You can do two things to spot the issue:
* Run your script with `CUDA_LAUNCH_BLOCKING=1` (which disables asynchronous CUDA execution)
* Make use of [`torch.autograd.detect_anomaly`](https://pytorch.org/docs/stable/autograd.html#torch.autograd.detect_anomaly) to detect the causes of NaNs
